Possible fix reimplementation for #19, by setting self.material = null

mikael-ros opened the pull request:
This should fix the issue in a minimal way. Hopefully it does not break other functionality.

mikael-ros commented:
I have now attempted to fix the issue I introduced. The thinking now is that upon reentry in the editor, the material is set back to the original material. 

Since ``_process`` is supposedly only called in the editor, I figured that'd be the best place to do so. To prevent infinite reapplications, the script checks if ``self.material`` is ``null`` whilst ``display_tilemap`` isn't (which would indicate that ``self.material`` wasn't ``null`` before.

It is possible that this has introduced a new bug, where removing a material no longer works. I'm quickly going to investigate such now

Another proposed solution would be to somehow instantaneously propagate any material changes to the display tile map, rather than juggling them around.

mikael-ros commented:
It did, in fact, turn out to be an issue. I have fixed it by also removing the display tilemaps material when the material is reapplied. 

Possible issues may remain with this implementation, however:
- if someone desires to add the material during runtime, it may propagate that setting to the editor. I don't know how plausible of an occurance that is though.

mikael-ros commented:
I had to force push to amend a typo in the commit messages. As you can see, I failed twice, third time was the charm... lol

mikael-ros commented:
Do note by the way, the same issue that causes the ``self.material = null`` to propagate to the editor also seems to apply for ``self.modulate`` and ``self.self_modulate``. Might want to look into that later, if it's important

mikael-ros commented:
It does break the ability to preview shaders in the editor, however, it seems.

pablogila reviewed (changes requested):
Thanks for reaching out!

Well, you said it. Neither I think the _process approach is ideal. This is a weird problem that I am not sure how to approach...

I am afraid that my knowledge in shaders is quite limited. Is there another way to hide the TileMapDual in the first place?

If there is not, Could we store the material as a resource file, so that Godot saves it even after closing the editor? But then, how can we access and modify that resource?

Maybe the exported variable is a better approach? Could we use only one variable instead of the messy two from the previous commits?

I would like to answer these questions to better approach this problem, so let's see...

Thanks a lot for sharing your ideas!

bendy1234 commented:
> Is there another way to hide the TileMapDual in the first place?

there are a few ways however all of them have downsides / changes you'd might not want to make
(i did not list every way to do this)
1. changing the visible flag, this disables editing in godot + visibility of child nodes, will need add the world tilemap as a sibling instead of a child
2. set `enabled` to false, however disables editing the tilemap in the editor (I do this at runtime)
3. changing modulate, shaders can modify modulate so they need to be set to null

you can also show/let the user add the tilemap and move it around. Just add `@export` to `display_tilemap` and change the `get_node_or_null("WorldTileMap")` part of the code
(^ this is what i did)

> Could we store the material as a resource file, so that Godot saves it even after closing the editor? But then, how can we access and modify that resource?

should be possible, probably will be messy. if you want to edit it in the inspector your better off not saving it as a file.

> Maybe the exported variable is a better approach? Could we use only one variable instead of the messy two from the previous commits?

this is probably the best approach that dosent change much but some people might get confused why there are multiple material properties in the inspector.  
you can add the material to the inspector like this without having the 2 separate `ShaderMaterial` and  `CanvasItemMaterial` fields
```
@export_custom(PROPERTY_HINT_RESOURCE_TYPE, "ShaderMaterial,CanvasItemMaterial")
var _material: Material = null
```

mikael-ros commented:
Instead of setting the material to ``null``, I have now made a version that makes the ``display_tilemap`` a child of whatever the ``TileMapDual`` is a parent of. This way, any visibility changes on ``TileMapDual`` do not affect the displayed tilemap, and edits that propagate to the editor are minimized.

You will immediately notice the ugly ``properties_match`` method. I was momentarily eyeing up a solution like the following for copying the properties: 

```gdscript
func copy_properties(source: Object, target: Object) -> void:
    for property in source.get_property_list():
        if property.name in target:
            target.set(property.name, source.get(property.name))
```

but it copied properties that made the layer a node of it's own in the tree permanently. A modification of this might work though. Similar logic applies to checking if properties have changed. 

A solution to this might be to reference an array of properties we want to change rather than copying all of them, but I'm not sure.

``properties_match`` is used to propagate any editor changes to the displayed tilemap.

It is possible that editing of material properties indirectly may not propagate still. This is so far still a suggestion and not a full on fix, I believe.

I do have to clarify I'm fairly new to Godot and GDScript. I'm more versed in other types of development. Sorry for any mistakes :O Criticism welcome

bendy1234 commented:
the current way of are adding the `display_tilemap` node will limit the number of `TileMapDual`s per node (and break others with the same parent) as the sibling is always called "WorldTileMap"  
Godot will automatically generate a name if you dont set it  

Also I dont think there is much of a performance problem with not checking if the properties are the same and just copying them over

mikael-ros commented:
Yeah, I figured it might. Perhaps creating a dummy node to parent them during runtime / rendering could work?

As far as performance is concerned, you're probably right.

I'll take a look at it again tonight :)

pablogila reviewed (commented):
Thanks for sharing. Just asking, why should we check if the properties match? Is there any reason we can't just copy them over?

mikael-ros commented:
It was mainly just because I momentarily thought it more performant to do so, but I realized later, like bendy, that it's probably almost just as performant to not check :o

mikael-ros commented:
I have now removed the checking, and parented the tile map layers with dummy nodes. 

This solves part of the issues, but is perhaps still not ideal

mikael-ros commented:
I have yet to find such a solution. However, I did play around with having a common node to house the tile map layers within, so that there aren't as many (currently 3) nodes per dual tile map. I ran into concurrency errors and race conditions though, I believe, which made it a lot more difficult to deal with (I am not very well versed with how multi-threading and task delegation works in the Godot engine and GDScript)

For now I have cleaned up the current solution slightly

pablogila reviewed (commented):
I am afraid this dummy approach is too complex. We should probably go back to an exported variable...

mikael-ros commented:
That is entirely okay, I was just exploring other avenues :)

pablogila commented:
I feel like one in these two approaches are the most reasonable:

> you can also show/let the user add the tilemap and move it around. Just add `@export` to `display_tilemap` and change the `get_node_or_null("WorldTileMap")`. this gives the user full control over the display tilemap (^ this is what i did)

This is how TileMapDual started. However it was a bit confusing for newcomers...

About the other approach,

> you can add the material to the inspector like this without having the 2 separate `ShaderMaterial` and `CanvasItemMaterial` fields. this method dose save the material even after godot is closed or restarted
> ```
> @export_custom(PROPERTY_HINT_RESOURCE_TYPE, "ShaderMaterial,CanvasItemMaterial")
> var _material: Material = null
> ```

This might be the least disruptive approach. Since we are only modifying the material to make TileMapDual transparent, we would only need an exported material. The rest of the characteristics can be copied from TileMapDual itself.

I feel like this last option might be the simplest one, and should not cause any unexpected bugs (fingers crossed).
What do you think? If I am missing something, please let me know.

mikael-ros commented:
Yes, I think the second option would be best too from a technical standpoint

Theres definetely two schools of thought, though. The second option is a lot easier to implement (as has been showed previously), but might be confusing for users. That's why I wanted to explore the possibility of doing it in a way where the normal material slot is used instead, but I personally don't mind the less intuitive approach - especially considering anything else has kind of been a bowl of spaghetti to implement, lol

I think, until Godot adds some reliable way of hiding the parent of a child without hiding the child or making it uneditable, or doing funky stuff in runtime, the second approach will be the cleanest. It would just entail some kind of note in the readme :)

I personally suggest you either close this PR and implement that version, or I can do so tomorrow. For me, this has mostly just been a learning experience, so I really don't mind having spent the time. I know a lot more about this addon now, shaders, and visibility within the Godot engine than I did before

pablogila commented:
Great! Of course, we are all here to learn! So, the exported material it is.
I don't have my laptop with me, so if you want to pr it you are more than welcome to do so :D
Otherwise I will do it next week or so. Well, whenever my laptop gets repaired, to be fair.
Cheers!

mikael-ros commented:
I'll take a look at it tomorrow then :)

What would you prefer?

1. I close this PR and open a new one with relevant changes, and move some of the optimizations over (cleaner PR, but loss of discussion and context)
2. I squash all prior commits in this PR into one commit and add new changes on top as a second commit (cleaner commit history)
3. I simply add new edits on top as new commits (simplest)


Hope your laptop gets fixed soon! :)

pablogila commented:
Go for 3 if you find it easier :D
Thanks!

mikael-ros commented:
I have now implemented the compromise. 

Additionally, I have updated the copying behaviour. Now, properties are only copied upon property change. To implement this for the propagated material, I have defined a custom setter. The reason for this is because I did notice some performance drawbacks / latency in addition to occasional difficulty setting certain properties (properties that are nullified don't always propagate due to concurrency errors, I think)

When the ordinary material slot is modified, the changes will be propagated upon scene reload (I couldn't get it to fire upon material change just yet, for some reason). My rationale is that an unknowing issue will then organically notice the way materials are propagated, rather than being frustrated and confused about why their material isn't being applied. In an ideal world, they would have read the readme, but it's just to prevent someone wasting time on troubleshooting it unknowingly. 

I have also updated the readme with this new behaviour described

Let me know if you prefer "Propagated material" or just "Material", or perhaps "Displayed Tilemap Material". I believe "Propagated Material" is the most concise and clear name, as "Material" is confusing, and "Displayed Tilemap Material" is a bit too verbose

pablogila reviewed (changes requested):
This is great!
Last thing on the variable naming: `display_material` seems shorter and concise, I hope it is clear for most users. Please let me know what you think and hit the button :D
I think I can later implement a setget so that the material gets reloaded automatically, without having to restart the scene.
Cheers!

mikael-ros commented:
I have updated it to be ``display_material`` now :)

pablogila reviewed (approved):
great!
